Fix exception handling

vincent-prz opened the pull request:
This aims to fix https://github.com/pybee/voc/issues/255

Does this look good?

vincent-prz committed:
Handle expressions in exception handler, WIP

vincent-prz committed:
Finish catching dynamic exceptions

eliasdorneles commented:
So I've been pairing on this with @Heisenberg815 for some time, I think we're pretty close now.

@Heisenberg815: i was looking at the build failure, it looks like we missed the `except (exc1, exc2)` case =)
To be continued !

BPYap commented on voc/python/ast.py:
```diff
@@ -2300,7 +2300,120 @@ def visit_Index(self, node):
+            "Warning",
+            "ZeroDivisionError",
+        ]
+        if node.type in builtin_exceptions:
```
@Heisenberg815 @eliasdorneles, so I was compiling asyncio modules using the codes from this PR and I found a minor bug where voc transpilation fails when `node.type` is None. 

A potential fix would be adding `or node.type == None` condition on this line to redirect the program flow to `self._visitCoreException(node)`.

Cheers 😃

vincent-prz committed:
Exception handler: add some fixes.

Handle the case where node.type is None.
Fix check for explicitely named builtin errors.

Heisenberg815 commented:
@BPYap I have made the fix you suggested, plus one other small fix. It seems I still have one test to fix, will do it shortly.

Sorry for the late response.

Cheers!

vincent-prz committed:
Merge branch 'master' into fix-exception-handling

vincent-prz committed:
Fix for custom exception in local namespace

vincent-prz committed:
Fix style

vincent-prz committed:
Add expected failure test for except with custom exception in tuple

eliasdorneles commented on tests/structures/test_try_catch.py:
```diff
@@ -912,3 +912,78 @@ def test_try_multiple_except_else_finally(self):
+            print('Done.')
+         """)
+
+        self.assertCodeExecution("""
```
Can you please put this test in a separate method, so that it can run as well?

This seems to be a regression (i.e., if i'm not mistaken, this second `assertCodeExecution` passes on master)

vincent-prz replied:
Done. Indeed, there seems to be a regression on `master`. But it seems it's the other way around: the second `assertCodeExecution` passes, while the first fails.

Do you think it's blocking the merge?

eliasdorneles replied:
Thanks!
Yeah, that's what I said -- the second passes. :)

So, while the patch seems to indeed already make things better, I would like to have a look with more patience to see what's causing that regression.
I'm a bit swamped lately but I'll have some days off soon and will look into it!

eliasdorneles commented on voc/python/ast.py:
```diff
@@ -2474,9 +2473,128 @@ def visit_ExtSlice(self, node):
+        )
+
     @node_visitor
+    # [FIXME] - make it work for Tuples
```
leftover to remove?

eliasdorneles reviewed (changes requested):
This is almost good, some minor changes before it's good!

vincent-prz committed:
Split expected failure tests in test_try_catch

vincent-prz committed:
Micro cleanup

vincent-prz committed:
Merge branch 'master' into fix-exception-handling

phildini commented:
Hi there! It looks like this PR might be dead, so we're closing it for now. Feel free to re-open it if you'd like to continue, or think about directing your efforts to https://github.com/beeware/briefcase or https://github.com/beeware/toga. Both of these have more active development right now. 😄
